src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexInterpreter.cs

```diff
@@ -804,15 +806,18 @@ private bool TryMatchAtCurrentPosition(ReadOnlySpan<char> inputSpan)
                         continue;
 
                     case RegexOpcode.EndZ:
-                        if (Rightchars() > 1 || Rightchars() == 1 && inputSpan[runtextpos] != '\n')
                         {
-                            break;
+                            int runtextpos = this.runtextpos;
+                            if (runtextpos < inputSpan.Length - 1 || ((uint)runtextpos < (uint)inputSpan.Length && inputSpan[runtextpos] != '\n'))
```

Comment:
Not following this, it previously checked `Rightchars() == 1` and now you're just checking it's at least 1?

Reply:
> it previously checked Rightchars() == 1

Can you clarify? It previously checked:
```
Rightchars() > 1 || (Rightchars() == 1 && inputSpan[runtextpos] != '\n')
```

Reply:
Oh, you mean my change on the second clause?  By the time we get to the second clause, we know that runtextpos >= inputSpan.Length - 1.  If runtextpos < inputSpan.Length is true, the only possible value runtextpos can be is inputSpan.Length - 1.  And checking for it this way then allows the JIT to prove that runtextpos is in the bounds of the span and eliminate the bounds check on the subsequent inputSpan[runtextpos].